src/ray/raylet/node_manager.cc

```diff
@@ -1022,17 +1071,49 @@ void NodeManager::ProcessRegisterClientRequestMessage(
     // Register the new driver.
     RAY_CHECK(pid >= 0);
     worker->SetProcess(Process::FromPid(pid));
+
     // Compute a dummy driver task id from a given driver.
     const TaskID driver_task_id = TaskID::ComputeDriverTaskId(worker_id);
     worker->AssignTaskId(driver_task_id);
     rpc::JobConfig job_config;
     job_config.ParseFromString(message->serialized_job_config()->str());
-    Status status = worker_pool_.RegisterDriver(worker, job_config, send_reply_callback);
-    if (status.ok()) {
-      auto job_data_ptr =
-          gcs::CreateJobTableData(job_id, /*is_dead*/ false, std::time(nullptr),
-                                  worker_ip_address, pid, job_config);
-      RAY_CHECK_OK(gcs_client_->Jobs().AsyncAdd(job_data_ptr, nullptr));
+
+    std::shared_ptr<rpc::JobTableData> job_table_data;
+    job_table_data = job_manager_.GetJobData(job_id);
+    if (job_table_data == nullptr) {
+      job_table_data = std::make_shared<JobTableData>();
+    }
+    job_table_data->set_job_id(job_id.Binary());
+    job_table_data->set_is_dead(false);
+    job_table_data->set_raylet_id(self_node_id_.Binary());
+    job_table_data->set_driver_ip_address(worker_ip_address);
+    job_table_data->set_driver_hostname(string_from_flatbuf(*message->hostname()));
+    job_table_data->set_driver_pid(pid);
+    job_table_data->set_language(language);
+    job_table_data->set_timestamp(current_time_ms());
+    *job_table_data->mutable_config() = job_config;
+
+    auto cb_register_driver = [this, send_reply_callback, job_table_data](
+                                  Status status, int assigned_port) {
+      if (!status.ok()) {
+        send_reply_callback(status, /*port=*/0);
+        return;
+      }
+
+      // Register job to GCS.
+      RAY_CHECK_OK(gcs_client_->Jobs().AsyncAdd(
+          job_table_data, [send_reply_callback, assigned_port](const Status &status) {
+            if (status.ok()) {
+              send_reply_callback(status, assigned_port);
+            } else {
+              send_reply_callback(status, /*port=*/0);
+            }
+          }));
+    };
+
+    Status status = worker_pool_.RegisterDriver(worker, job_config, cb_register_driver);
+    if (!status.ok()) {
+      send_reply_callback(status, /*port=*/0);
```

Comment:
same here, should always call the callback in `RegisterDriver`

Reply:
I made the `WorkerPool::RegisterDriver` returns void, so the status can only be retrieved from the callback.